// ngx-echarts/esm5/lib/change-filter.js
import { of, EMPTY } from 'rxjs';
var ChangeFilter = /** @class */ (function () {
    function ChangeFilter(changes) {
        this.changes = changes;
    }
    ChangeFilter.of = function (changes) {
        return new ChangeFilter(changes);
    };
    ChangeFilter.prototype.notEmpty = function (key) {
        if (this.changes[key]) {
            var value = this.changes[key].currentValue;
            if (value !== undefined && value !== null) {
                return of(value);
            }
        }
        return EMPTY;
    };
    ChangeFilter.prototype.has = function (key) {
        if (this.changes[key]) {
            var value = this.changes[key].currentValue;
            return of(value);
        }
        return EMPTY;
    };
    ChangeFilter.prototype.notFirst = function (key) {
        if (this.changes[key] && !this.changes[key].isFirstChange()) {
            var value = this.changes[key].currentValue;
            return of(value);
        }
        return EMPTY;
    };
    ChangeFilter.prototype.notFirstAndEmpty = function (key) {
        if (this.changes[key] && !this.changes[key].isFirstChange()) {
            var value = this.changes[key].currentValue;
            if (value !== undefined && value !== null) {
                return of(value);
            }
        }
        return EMPTY;
    };
    return ChangeFilter;
}());
export { ChangeFilter };
